### PR TITLE
Perf(common): optimize cache promotion and demotion

### PR DESCRIPTION
**Pull request by zztaki**
### What problem does this PR solve?

Issue Number: #2946 

Problem Summary: current cache promotion and demotion need to copy a duplica, erase old elem and reinsert duplica.

### What is changed and how it works?

What's Changed: optimize cache promotion and demotion using `list.splice` and add some tests.

How it Works: `list.splice` will update cache object priority without object duplica, list reinsertion and hashtable update.

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [x] Relevant documentation/comments is changed or added
- [x] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
This is a performance test.
```cpp
#include <benchmark/benchmark.h>
static void GetWithDuplicate(benchmark::State& state){
    auto cache = std::make_shared<LRUCache<int, int>>(5, nullptr);
    for (int i = 0; i < 5; i++) {
        cache->Put(i, i);
    }
    int v;
    for (auto _ : state) {
        for (int k = 0; k < 5; k++) {
            cache->Get(k, &v);
            assert(k == v);
        }
    }
}
BENCHMARK(GetWithDuplicate);

static void GetWithoutDuplicate(benchmark::State& state){
    auto cache = std::make_shared<LRUCache<int, int>>(5, nullptr);
    for (int i = 0; i < 5; i++) {
        cache->Put(i, i);
    }
    int v;
    for (auto _ : state) {
        for (int k = 0; k < 5; k++) {
            cache->GetWithSplice(k, &v);
            assert(k == v);
        }
    }
}
BENCHMARK(GetWithoutDuplicate);

BENCHMARK_MAIN();
```
Result is
image

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
@Ziy1-Tan Hello, this is my first encounter with CI. In fact, I've added some relevant tests that should completely cover my modifications. But CI tells me coverage ratio 72.9 < expectd 73. I don’t know if this is because my modifications are not up to standard or if it’s due to other modules. I hope I can get your help. :)

**Comment by zztaki** [comment]
Maybe I can add some tests for `arc_cache`.

File: src/common/lru_cache.h [hunk 1]
```diff
@@ -699,16 +696,8 @@ bool SglLRUCache<K, KeyTraits>::MoveBack(const K &key) {
     if (iter == cache_.end()) {
         return false;
     }
-    // delete the old value
-    RemoveElement(iter->second);
-    // put new value at tail
-    ll_.push_back(key);
-    cache_[key] = --ll_.end();
-    size_++;
-    if (cacheMetrics_ != nullptr) {
```
Comment (Ziy1-Tan) [inline src/common/lru_cache.h:708]: Why delete these lines?

File: src/common/lru_cache.h [hunk 2]
```diff
@@ -699,16 +696,8 @@ bool SglLRUCache<K, KeyTraits>::MoveBack(const K &key) {
     if (iter == cache_.end()) {
         return false;
     }
-    // delete the old value
-    RemoveElement(iter->second);
```
Comment (Ziy1-Tan) [inline src/common/lru_cache.h:703]: This function also include some metrics code, maybe you should consider it?

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by Ziy1-Tan** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
cicheck

**Comment by zztaki** [comment]
Since the line coverage and branch coverage of the existing `src/client` are not up to standard, an additional commit which adds some tests was submitted.

**Comment by zztaki** [comment]
close #2982

**Review by Ziy1-Tan — APPROVED** [review]

**Review by wuhongsong — APPROVED** [review]

**Pull request merged**